### ISSUE TITLE
Context Data Update Behavior

### ISSUE THREAD

**Issue by matthias-pichler-warrify** [serverlessworkflow/specification#867]
**What would you like to be added**:

The `output.to` field allows to write data to the context. Is the context merged or overwritten by the data produced by this expression?

given a context of

```json
{
  "foo": "bar"
}
```

will the expression

```yaml
output:
  to: '{a: 1}'
```

result in

```json
{
  "foo": "bar",
  "a": 1
}
```

or 

```json
{
  "a": 1
}
```

Personally always extending makes sense to me from a usability perspective but it seems to me that context could really explode in size this way. Especially because jq has built in methods for extending objects, such as

```yaml
output:
  to: '.a = 1'
```

or

```yaml
output:
  to: '$context + { a: 1 }'
```

which both are more explicit about extending

**Why is this needed**:

This is relevant for making sure data flow is implemented correctly and context does not explode in size.

**Comment by cdavernas** [comment]
Well, yet another excellent question 😆 

My opinion on this is that the result should be overwritten. The reason is that, as you said, `jq` provides ways to replace, update and extend. 

I feel that if we just extend, we would throw down the bin some very cool `jq` features, would "force" runtimes to perform an additional processing after the evaluation and would probably loose some portability accross expression evaluation languages (in Synapse, we allow using both `jq` and `js`, for example). Additionally, as you also said, the context could explode in size.

**Comment by fjtirado** [comment]
I think the problem is the ambiguity of to. 
Why not adding "merge" and "set", if using the former, whatever json generated by the expression after "merge" is merged into the context, while is using "set", the context will be replaced by the json generated by the expression after "set".

**Comment by cdavernas** [comment]
To is ambiguous indeed. However it is not solved imho by using merge or set, as it's the content of the jq expression which (should) decide what to do (just update, replace or merge)

**Comment by matthias-pichler-warrify** [comment]
> To is ambiguous indeed. However it is not solved imho by using merge or set, as it's the content of the jq expression which (should) decide what to do (just update, replace or merge).

I agree. I don't think `merge` or `set` solve the problem. 

But I also think there is a problem with relying on jq expressions to handle updates (this is my understanding of how jq works, so correct me if I am wrong): jq can only operate in the context of one object.

given:

```json
{
  "context": {
    "foo": "bar"
  },
  "output": {
    "a": 1
  }
}
```

and the following output expression:

```yaml
output:
  to: '.b = [$.context.foo, .a]'
```

The desired output would be:

```json
{
  "foo": "bar",
  "b": ["bar", 1]
}
```

If this expression is now run in the context of `context` it would update `context` to be:

```json
{
  "foo": "bar",
  "b": ["bar", null]
}
```

because `.a` cannot be resolved relative to `context`. If however the expression is run in context of the `output` it would result in:

```json
{
  "a": 1,
  "b": ["bar", 1]
}
```

which is the modified `output` not `context`.

Because of that I think it would make the most sense if:

1. the expression is run in context of the `output` (because then accessing the output and `$context` is possible)
2. the result of the expression _replaces_ the context (this avoids size explosions)

a potential rename of the `output.to` might help or just explicit documentation.

**Comment by cdavernas** [comment]
>  jq can only operate in the context of one object.

Yes, that is correct. The operand being the "input".

... but you have parameters, which brings me to the `$output` variable, which is however not documented by mistake. This variable should only be present when evaluating output expressions.

**Comment by cdavernas** [comment]
> (I propose to use set). And on top of that I propose to add merge

That's an intersting idea indeed. I feel however it adds additional properties for something that is already explicitly written by the user, which is the expression. If the user writes an expression that overwrites the context, so be it: if she's smart enough to write the expression, let's assume she's smart enough to understand it ^^.

If we choose to rename `to`, instead of `set`, I propose we consider `output` as a verb, and find an alertenative that "flows", which is why we chose the `to` term in the first place.

I'm not even sure there's a need to rename at all, the issue as I understand it being more about explicitly settling on/documenting the behavior.

**Comment by fjtirado** [comment]
@cdavernas  
Lets rewing and try to write in english what we are proposing to do. 
We are setting the context of the workflow with the value result of the expression. 
I would say that proper syntax should be 

`output: <expr> to: <variable to be set>`

**Comment by cdavernas** [comment]
> I would say that proper syntax should be

Renaming is not the point here: it's off topic. If that's something you think it's necessary, let's open an issue or a discussion about it 😉 

> output: <expr> to: <variable to be set>

As said in my previous commentary, I see no need to "split" the functionality of the expression feature into two different variables. Furthermore, with proposed syntax, you left out the output schema. 
As usual, those changes are subtles yet very impactful, and require a proper dedicated discussion in my personal opinion.

**Comment by fjtirado** [comment]
How is this a different discussion if we are just discussing how to update the context?

**Comment by fjtirado** [comment]
https://github.com/serverlessworkflow/specification/issues/873

**Comment by cdavernas** [comment]
@matthias-pichler-warrify Hasn't this been addressed and closed by #873 ?

**Comment by cdavernas** [comment]
@matthias-pichler-warrify can we close this issue?
